### PR TITLE
Track caller's emitted events

### PR DESCRIPTION
**Pull request by sug0**
## Describe your changes

Track the source code origin of some emitted event in Namada.

## Indicate on which release or other PRs this topic is based on

`v0.36.1`

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: crates/events/src/tracer.rs [hunk 1]
```diff
@@ -227,6 +228,29 @@ mod tests {
...
+            emit_event(ev, &mut events);
+            events.pop().unwrap()
+        };
+
+        let trace = event.read_attribute::<EventOrigin>().unwrap();
+
+        assert!(
+            trace.line > _EMIT_EVENT_START_LINE
+                && trace.line < _EMIT_EVENT_END_LINE
+        );
```
Comment (sug0) [inline crates/events/src/tracer.rs:248]: just realized this is wrong :(

we must check that the event originated...

File: crates/events/src/tracer.rs [hunk 2]
```diff
@@ -227,6 +228,29 @@ mod tests {
...
 
+    #[test]
+    fn test_event_trace_emit_event() {
+        let event = {
+            let ev = Event::new(
+                EventTypeBuilder::new_with_type("test").build(),
+                EventLevel::Tx,
+            );
+            let mut events = Vec::with_capacity(1);
+            emit_event(ev, &mut events);
```
Comment (sug0) [inline crates/events/src/tracer.rs:239]: ...here instead

Comment (tzemanovic) [reply crates/events/src/tracer.rs:239]: it's right, no? I think it's more like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d5cf2e4b13d8cd82d7d8046657a9243b

File: crates/events/Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,7 @@ version.workspace = true
 
 [features]
 default = []
+debug = []
```
Comment (tzemanovic) [inline crates/events/Cargo.toml:17]: we can probably add this to `testing = ["debug"]`

**Review by tzemanovic — APPROVED** [review]

File: crates/events/src/tracer.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            const START_LINE: u32 = line!();
+            emit_event(ev, &mut events);
+            const END_LINE: u32 = line!();
+
+            (events.pop().unwrap(), START_LINE, END_LINE)
+        };
+
+        let trace = event.read_attribute::<EventOrigin>().unwrap();
+
+        assert!(trace.line > start_line && trace.line < end_line);
```
Comment (brentstone) [inline crates/events/src/tracer.rs:255]: nit: maybe `trace.line == start_line + 1 && trace.line == END_LINE - 1`?

Comment (sug0) [reply crates/events/src/tracer.rs:256]: doesn't matter to be precise, we only care to check if the event was emitted within a certain region of the source code

**Pull request merged**